src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -0,0 +1,591 @@
+
+				// add the bytes read from the pointer to the position
+				position += bytesRead;
+
+				// remove the current primary reference if the current position has reached the end of the bytes
+				if (d.Position == d.Length) {
+					lock (dataLock) {
+						// this is the same object, it was done to make the cleanup 
+						data.Dequeue ();
+						current?.Dispose ();
+						currentStream?.Dispose ();
+						current = null;
+						currentStream = null;
+					}
+				}
+
+				return bytesRead;
+			}
+
+			public override bool CanRead => true;
+
+			public override bool CanSeek => false;
+
+			public override bool CanWrite => false;
+
+			public override bool CanTimeout => false;
+
+			public override long Length => length;
+
+			public override void SetLength (long value)
+			{
+				throw new InvalidOperationException ();
+			}
+
+			public override long Position {
+				get { return position; }
+				set { throw new InvalidOperationException (); }
+			}
+
+			public override long Seek (long offset, SeekOrigin origin)
+			{
+				throw new InvalidOperationException ();
+			}
+
+			public override void Flush ()
+			{
+				throw new InvalidOperationException ();
+			}
+
+			public override void Write (byte [] buffer, int offset, int count)
+			{
+				throw new InvalidOperationException ();
+			}
+		}
+
+#if MONOMAC
+		// Needed since we strip during linking since we're inside a product assembly.
+		[Preserve (AllMembers = true)]
+#endif
+		private class WrappedNSInputStream : NSInputStream
+		{
+			private NSStreamStatus _status;
+			private readonly Stream _stream;
+
+			public WrappedNSInputStream (Stream stream)
+			{
+				_status = NSStreamStatus.NotOpen;
+				_stream = stream;
+			}
+
+			public override NSStreamStatus Status {
+				get {
+					return _status;
+				}
+			}
+
+			public override void Open ()
+			{
+				_status = NSStreamStatus.Open;
+				Notify (CFStreamEventType.OpenCompleted);
+			}
+
+			public override void Close ()
+			{
+				_status = NSStreamStatus.Closed;
+			}
+
+			public override nint Read (IntPtr buffer, nuint len)
+			{
+				var source = new byte [len];
+				var read = _stream.Read (source, 0, (int)len);
+				Marshal.Copy (source, 0, buffer, (int)len);
+
+				//if (read == 0)
+				//    Notify(CFStreamEventType.EndEncountered);
+
+				return read;
+			}
+
+			public override bool HasBytesAvailable ()
+			{
+				return true;
+			}
+
+			protected override bool GetBuffer (out IntPtr buffer, out nuint len)
+			{
+				// Just call the base implemention (which will return false)
+				return base.GetBuffer (out buffer, out len);
+			}
+
+			protected override bool SetCFClientFlags (CFStreamEventType inFlags, IntPtr inCallback, IntPtr inContextPtr)
+			{
+				// Just call the base implementation, which knows how to handle everything.
+				return base.SetCFClientFlags (inFlags, inCallback, inContextPtr);
+			}
+
+			bool notifying;
+			[Export ("_scheduleInCFRunLoop:forMode:")]
```

Comment:
This is not public API, which means we can't use it (since it could cause apps using Xamarin.iOS to be rejected from the Apple Store).

This might be the reason the original ModernHttpClient implementation read the data into a byte array.